odoo/tools/translate.py

```diff
@@ -1150,52 +1209,136 @@ def trans_load_data(cr, fileobj, fileformat, lang,
-        # the entries from the POT file).
         for row in reader:
-            process_row(row)
+            if not row.get('value') or not row.get('src'):  # ignore empty translations
+                continue
+            if row.get('type') == 'code':  # ignore code translations
+                continue
+            # TODO: CWG if the po file should not be trusted, we need to check each model term
+            model_name = row.get('imd_model')
+            module_name = row['module']
+            imd_name = row['imd_name']
+            if model_name in env:
+                fname = row['name'].split(',')[1]
+                field = env[model_name]._fields[fname]
+                if not field.translate or not field.store:
+                    continue
+                xmlid = module_name + '.' + row['imd_name']
+                module_imd_names.append(module_name)
+                module_imd_names.append(imd_name)
+                model_name_to_field_name_to_xmlids[model_name][fname].append(xmlid)
+                translations[model_name][xmlid][fname][row['src']] = row['value']
+
+        # get xmlid to record id map
+        if not module_imd_names:
+            return
+        cr.flush()
+
+        # TODO: CWG check if split in another way _lookup_xmlids
+        # TODO: this approach deal with all logic in python, try to move model translation to SQL
+        xmlid_to_id = {}
+        for sub_module_imd_names in cr.split_for_in_conditions(module_imd_names, size=Cursor.IN_MAX * 2):
+            query = "SELECT module || '.' || name, res_id, noupdate, model FROM ir_model_data WHERE "
+            query += " OR ".join(["module = %s AND name = %s"] * int(len(sub_module_imd_names) / 2))
+            cr.execute(query, sub_module_imd_names)
+            rows = cr.fetchall()
+            # {xmlid: (id, noupdate, model_name)}
+            xmlid_to_id.update([(row[0], row[1:]) for row in rows])
+        # {(model_name, id): (xmlid, noupdate)}
+        id_to_xmlid = {(v[2], v[0]): (k, v[1]) for k, v in xmlid_to_id.items()}
+
+        # translate model terms
+        for model_name, model_dictionary in translations.items():
+            if model_name not in env:
+                continue
+            Model = env[model_name]
+            model_table = Model._table
+            field_name_to_xmlids = model_name_to_field_name_to_xmlids[model_name]
+            fields = Model._fields
+            field_type = defaultdict(list)
+            for field_name in field_name_to_xmlids.keys():
+                translate = fields[field_name].translate
+                field_type['no_translate' if not translate else 'model_term' if callable(translate) else 'model'].append(field_name)
+            if field_type['no_translate']:
+                pass  # ignore illegal translations
+            for field_name in field_type['model_term']:
+                record_ids = set(xmlid_to_id.get(xmlid, (0, False, ''))[0] for xmlid in field_name_to_xmlids[field_name])
```

Comment:
Does it really make sense to keep the id 0 around if we don't find a match?
```suggestion
                record_ids = {
                	it[0]
                	for it in map(xmlid_to_id.get, field_name_to_xmlids[field_name])
                	if it
                }
```
should do the trick no?